export doc (#1849)
arbor-sim/arbor: doc/internals/export.rst

based on the build variant, compiler and platform.

By default, hidden/visible symbols will affect **shared** libraries directly. Since the linker is
responsible for making symbols hidden, the visibility settings are not applied to **static**
libraries immediately, as static libraries are generated by bundling object files with the archiver
(standard in cmake for ``STATIC`` libraries).  In principle, the linker would be able to generate
relocatable output instead (``ld -r``) by incrementally linking object files into one output file
(sometimes called partial linking) which would apply proper visibility to static libraries.
However, we currently do not handle this case in our build scripts as it is not not nativley
`supported by cmake yet <https://gitlab.kitware.com/cmake/cmake/-/issues/16977>`_.

.. Note::
    When linking an application with **static** Arbor libraries the linker may issue warnings (particularly on macos).  Thus, if you encounter problems, try building shared Arbor libraries (cmake option ``-DBUILD_SHARED_LIBS=ON``) instead.

Macro Descripiton
-----------------

.. c:macro:: ARB_LIBNAME_API

>     Here ``LIBNAME`` is a placeholder for the library's name: ``ARB_ARBOR_API`` for the core Arbor

Reviewer: ```suggestion
    Here "``LIBNAME``" is a placeholder for the library's name: ``ARB_ARBOR_API`` for the core Arbor
```
Think you accidentally deleted them.

    library, ``ARB_ARBORIO_API`` for Arborio, etc. This macro is intended to annotate functions,
    classes and structs which need to be accessible when interfacing with the library. Note that it
    expands to different values when Arbor is being built vs. when Arbor is being used by an
    application.
    
    Below we list the places where the macro needs to be added or can be safely omitted (we assume
    all of the symbols below are part of the public/user-facing API).  In general, annotation is
    required
    
    * for **declarations in header files** which are not definitions

>     * for **defintions** of functions, friend functions and (extern) variables **in source files**

Reviewer: ```suggestion
    * for **definitions** of functions, friend functions and (extern) variables **in source files**
```

        
    Members and member functions of already visible classes will be visible, as well, without
    further annotation, except for friend function (see below).

    Implementation details and internal APIs may not need annotation as long as they do not require
    visibility across the library boundary (though some annotations are required for unit test
    purposes). Exception classes and type-erased objects need special annotation, see
    :c:macro:`ARB_SYMBOL_VISIBLE`.

    .. code-block:: cpp
        :caption: header.hpp

        #include <arbor/export.hpp>

        // free function declaration
        ARB_ARBOR_API void foo();

        // free function definition
        inline void bar(int i) { /* ... */ }

        // function template definition
        template<typename T>
        void baz(T i) { /* ... */ }

        // class definition
        // class member declaration
        class ARB_ARBOR_API A {
            A();
            friend std::ostream& operator<<(std::ostream& o, A const & a);
        };

>         // class defintion

Reviewer: ```suggestion
        // class definition
```

>         // class member defintion

Reviewer: ```suggestion
        // class member definition
```
:)

        class B {
            B() { /* ... */ }
        };

>         // template class defintion

Reviewer: ```suggestion
        // template class definition
```

        // class member definition
        template<typename T>
        class C {
            C() { /* ... */ }
        };

        // (extern) variable declarations
        ARB_ARBOR_API int g;
        ARB_ARBOR_API extern int h;

    .. code-block:: cpp
        :caption: source.cpp

        // free function definition
        ARB_ARBOR_API void foo() { /* ... */ }

>         // class member defintion (will be visible since A is visible)

Reviewer: ```suggestion
        // class member definition (will be visible since A is visible)
```

        A::A() { /* ... */ }

>         // friend function defintion

Reviewer: ```suggestion
        // friend function definition
```

.. _export:

Exporting Symbols
=================

The arbor libraries are compiled with hidden visibility by default. Thus, symbols which are part of

> the public API need to be marked explicitely as visible. Arbor provides a couple of macros to

Reviewer: ```suggestion
the public API need to be marked explicitly as visible. Arbor provides a couple of macros to
```

annotate functions and classes which are defined in the header file ``export.hpp`` in each library's
include directory, i.e. ``include/arbor/export.hpp``. These header files are generated at configure
time based on the build variant, compiler and platform.

Macro Descripiton
-----------------

.. c:macro:: ARB_LIBNAME_API

>     Here ``LIBNAME`` is a placeholder for the library's name: ``ARB_ARBOR_API`` for the main arbor

Reviewer: ```suggestion
    Here "``LIBNAME``" is a placeholder for the library's name: ``ARB_ARBOR_API`` for the core Arbor
```

>     library, ``ARB_ARBORIO_API`` for arborio, etc. This macro is intended to annotate functions,

Reviewer: ```suggestion
    library, ``ARB_ARBORIO_API`` for Arborio, etc. This macro is intended to annotate functions,
```

>     classes and structs which need to be accessible when using the library. Note, that it expands to

Reviewer: ```suggestion
    classes and structs which need to be accessible when interfacing with the library. Note that it expands to
```
What's 'it' in 'it expands'?

>     different values when arbor is being built vs. when arbor is being used by an application. Below

Reviewer: ```suggestion
    different values when Arbor is being built vs. when Arbor is being used by an application. Below
```

    we list the places where the macro needs to be added or can be safely omitted (we assume all of

>     the symbols below are part of the public API).

Reviewer: public = not hidden? Earlier you've used 'visible' as opposite for 'hidden'. Prefer consistency (I see the GCC page does not set a good example either).

Author: public in the sense of accessible (vs. private, implementation details). Public/private is an orthogonal concept to hidden/visible. Both public and private symbols are affected by visibility. I could replace "public API" with user-facing API, would that help? I don't think I've used public elsewhere to describe non-hidden - that would indeed be confusing.

Reviewer: It's clear now that I am not an expert. My knowledge is limited to private/public/protected properties/methods, which are sometimes described as visibility, and exported symbols (or symbols not exported, which you might describe as hidden).
It seems that hidden/visible has another specific meaning which I struggle to disentangle from what one learns in every first year C++ course.

Is it possible to have a line or two right at the beginning of the page that clarify the difference?

Author: done

Reviewer: Thanks!

    .. code-block:: cpp
        :caption: header.hpp

>         // free function declaration

Reviewer: What's a 'free function'?

Author: not a class member

Reviewer: OK!

        ARB_ARBOR_API void foo();

>         // free function (inline)

Reviewer: 'inline' refers to not being exposed/hidden? If so, prefer 'hidden'. If not, educate me :)

Author: inline means, the function body (the definition) of the function resides in the header (it is compiled anew for each usage of Arbor). Conversely non-inline functions are compiled once and are part of the pre-compiled library (only the declaration is in the header).

Reviewer: Of course, the absence of `ARB_ARBOR_API` ticked me off.

Then there are two differences with the 'free function': absence of the annotation and the presence of a body. Whether a free function is inline or not, does not affect how to annotate, right? If so, maybe leave it out.

Author: it does affect annotation:
- inline free function (definition in header): don't annotate (is visible no matter what)
- free function (declaration in header, definition in source): must annotate both declaration and definition (is hidden otherwise)

Reviewer: Can you put those bullets in the documentation? Very clear :)

Author: I did add 2 bullets in the text - let me know if that's ok

Reviewer: Yes!

        void bar(int i) { /* ... */ }

        // function template (inline)
        template<typename T>
        void baz(T i) { /* ... */ }

        // class declaration

>         // note: this will make all member symbols visible

Reviewer: member symbols will become visible as per class visibility directives or do you mean that those are overridden as well?

Author: I don't quite understand. All members of a class (member functions and members) will be implicitly annotated with hidden=default (i.e. visible) if the class is made visible through the macro.

Reviewer: ```
class B {
   int myint;
};
```
`myint` is hidden because class members are hidden by default. `B` is hidden from the public API, because it isn't annotated to be visible. Right?
```
class ARB_ARBOR_API B {
   int myint;
};
```
I would expect `myint` to be hidden, even though having annotated `B` to be visible. Is that so?

        // template class (inline)
        template<typename T>
        class C {
            /* ... */
        };

        // (extern) global variable declarations
        ARB_ARBOR_API int g;
        ARB_ARBOR_API extern int h;

    .. code-block:: cpp
        :caption: source.cpp

        // free function definition
        ARB_ARBOR_API void foo() { /* ... */ }

        // class member functions
        A::A() { /* ... */ }

        // friend functions
        ARB_ARBOR_API std::ostream& operator<<(std::ostream& o, A const& a) { /* ... */ }

        // (extern) global variable definitions
        ARB_ARBOR_API int g = 10;
        ARB_ARBOR_API int h = 11;

.. c:macro:: ARB_SYMBOL_VISIBLE

    Objects which are type-erased and passed across the library boundaries sometimes need runtime
    type information (rtti). In particular, exception classes and classes stored in ``std::any`` or
    similar need to have the correct runtime information attached. Hidden visibility strips away
    this information which leads to all kind of unexpected behaviour. Therefore, all such classes
    must be annotated with this macro which guarantees that the symbol is always visible. Note, it

>     is not enough to use the first macro for these cases.

Reviewer: To avoid confusion: when you say that the above macro isn't enough, do you mean that the following is an additional specifier or that this should be used instead? Looking at the example, probably instead, but the language could be ambiguous.

Author: yes I mean: should be used instead. Will clarify
